### ISSUE TITLE
[enhancement] add logger timestamp to terminal output, display logs in a uniform manner

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1367]
- [ ] maybe provide an option to turn off color
- [ ] if levels are displayed for error/warning, probably better to always display the levels
- [ ] display the timestamp in the stdout like it's done in the ASLOGS

I've been running the testing suite this week, and had a few different `nohup`'ed processes. Each of these are running a subset of the tests. One `nohup` may be from last week, the other from Monday this week, another from yesterday...

I noticed that it is quite impossible to know what was executed before/after what, by looking at the logs. This is bad for troubleshooting, maintaining experiments, and also traceability. For example, today `t01s` failed, and my nohup contained only this:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/980ea5a21f5d92b853e3f8aeeec07c02/image.png)

Note that other lines have only the log message. But this one has the log level, followed by the message. It's probably be better to always display the level, or not display the level at all (I think the former), so that our logs can be parsed (e.g. if someday we ingest them to produce a dashboard with something like Grafana, for instance, then it must adhere to some standard).

There are several `ESC[0m[ESC` lines in the nohup log (stdout of `autosubmit run` which I think are the terminal colors. While useful to users in interactive mode, when you `nohup` the process, and then open on `less` or `vim`, these are highlighted with the same pattern as when you search for text. If we allow users to use `--no-color`, for instance, then their editors will produce less noise, which helps when they are troubleshooting some experiment.

Finally, the ASLOGS do contain the timestamp for that. You just need to find the ASLOGS and be quite sure it's the same line you saw in your `nohup`'ed output (still safer if I can have the exact same line in both). For example:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/b967b5735a19b7a4323f99b377863e8f/image.png)

Note, that even though it shows the timestamp, there are WARNING/ERROR also only in certain lines. The others are probably INFO, which may sound like a good idea to reduce noise to users, but in reality may confuse if users are not aware that the level is what's bracketed, and again, it makes it harder to parse/grep/awk/etc log files.

TL;DR: we can make our logs more uniform, allowing them to be more easily parsed by humans and machines, and simplify finding out when something goes wrong

**Comment by kinow** [comment]
CWLTool shows only the level and message for brevity, but it allows you to turn on timestamps, giving you a standardized log output. It also allows you to `--disable-color`.

```bash
(venv) kinow@ranma:/tmp$ cwltool --timestamps --disable-color hello_world.cwl
[2024-07-19 08:51:25] INFO /tmp/venv/bin/cwltool 3.1.20240708091337
[2024-07-19 08:51:25] INFO Resolved 'hello_world.cwl' to 'file:///tmp/hello_world.cwl'
[2024-07-19 08:51:26] INFO [job hello_world.cwl] /tmp/7lvw_vml$ echo \
    'Hello World'
Hello World
[2024-07-19 08:51:26] INFO [job hello_world.cwl] completed success
{}[2024-07-19 08:51:26] INFO Final process status is success
```

Cylc too, you can turn on/off timestamps, and colors.

```bash
(cylc) kinow@ranma:~/cylc-src/five$ cylc play --timestamp --no-detach --color=never five/run4

 ▪ ■  Cylc Workflow Engine 8.3.2
 ██   Copyright (C) 2008-2024 NIWA
▝▘    & British Crown (Met Office) & Contributors

2024-07-19T09:01:06+02:00 INFO - Extracting job.sh to /home/kinow/cylc-run/five/run4/.service/etc/job.sh
2024-07-19T09:01:06+02:00 INFO - Workflow: five/run4
2024-07-19T09:01:06+02:00 INFO - Scheduler: url=tcp://ranma:43076 pid=10756
2024-07-19T09:01:06+02:00 INFO - Workflow publisher: url=tcp://ranma:43065
2024-07-19T09:01:06+02:00 INFO - Run: (re)start number=1, log rollover=1
2024-07-19T09:01:06+02:00 INFO - Cylc version: 8.3.2
2024-07-19T09:01:06+02:00 INFO - Run mode: live
2024-07-19T09:01:06+02:00 INFO - Initial point: 1
2024-07-19T09:01:06+02:00 INFO - Final point: None
2024-07-19T09:01:06+02:00 INFO - Cold start from 1
2024-07-19T09:01:06+02:00 INFO - New flow: 1 (original flow from 1) 2024-07-19T09:01:06
2024-07-19T09:01:06+02:00 INFO - [2/foo:waiting(runahead)] => waiting
2024-07-19T09:01:06+02:00 INFO - [2/foo:waiting] => waiting(queued)
2024-07-19T09:01:06+02:00 INFO - [3/foo:waiting(runahead)] => waiting
...
...
```

@ainagaya you are the next to run the testing suite, but you also have a lot of experience looking at these logs :rofl: so tagging you here just in case you have some other ideas to improve the logging output at some point :thumbsup:

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 19, 2024, 09:21

Hi @kinow , thanks for tagging me! I think it could be useful. I also wanted to improve the logs of the DestinE workflow, adding a logger... I'll open an issue for that, let's see when I can find some time to do it :smile:

**Comment by kinow** [comment]
In GitLab by @ainagaya on Jul 19, 2024, 09:24

mentioned in issue digital-twins/de_340-2/workflow#501

**Comment by kinow** [comment]
Good idea! And the applications also need a logger. I noticed this while writing the unit test where I saw `print` statements. I suspect most apps are doing the same, only throwing exceptions and printing. Thanks!
